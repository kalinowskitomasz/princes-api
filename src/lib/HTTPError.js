export default class HTTPError extends Error {
    constructor(message, status) {
        super(message)
        this.status = status
    }
}
