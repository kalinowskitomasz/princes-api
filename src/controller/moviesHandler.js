import getMovies from '../service/getMovies'

export default async function moviesHandler(params, event) {
    const init = {
        headers: { 'content-type': 'application/json;charset=UTF-8' },
    }

    const allMovies = await getMovies()

    const response = new Response(JSON.stringify(allMovies, null, 2), init)

    return response
}
