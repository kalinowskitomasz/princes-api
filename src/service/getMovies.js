import { StatusCodes } from 'http-status-codes'

import moviesHandler from '../controller/moviesHandler'
import HTTPError from '../lib/HTTPError'
import { getCinemaworldMovies } from '../repository/getCinemaworld'
import { getFilmworldMovies } from '../repository/getFilmworld'

function mergeDataSources(allMovies) {
    const moviesMap = new Map()
    allMovies.forEach(source => {
        source.data.forEach(movie => {
            // Take advantage of the fact that both services use
            // IMDB id with modified prefix
            // revert it back to original for consistency and data integrity
            const imdbId = `tt${movie.ID.slice(2)}`
            if (!moviesMap.has(imdbId)) {
                moviesMap.set(imdbId, { ...movie, ID: imdbId })
            }
        })
    })
    return [...moviesMap.values()]
}

export default async function getMovies() {
    const allMovies = await Promise.all([
        getCinemaworldMovies(),
        getFilmworldMovies(),
    ])

    // if at least one of the sources returns
    // response still could be sent
    let success = false
    allMovies.forEach(response => {
        success = success || !response.error
    })

    if (!success)
        throw new HTTPError('getMovies failed', StatusCodes.BAD_GATEWAY)

    const moviesData = mergeDataSources(allMovies)

    if (moviesData.length === 0)
        throw new HTTPError('no movies found', StatusCodes.NOT_FOUND)

    return moviesData
}
