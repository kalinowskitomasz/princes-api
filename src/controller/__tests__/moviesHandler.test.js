import moviesHandler from '../moviesHandler'
import getMovies from '../../service/getMovies'
import HTTPError from '../../lib/HTTPError'
import { StatusCodes } from 'http-status-codes'

jest.mock('../../service/getMovies')

describe('moviesHandler', () => {
    beforeEach(() => {
        getMovies.mockReset()
    })

    test('returns valid response', async () => {
        getMovies.mockResolvedValue(
            {
                ID: 'tt2488496',
                Poster:
                    'https://m.media-amazon.com/images/M/MV5BOTAzODEzNDAzMl5BMl5BanBnXkFtZTgwMDU1MTgzNzE@._V1_SX300.jpg',
                Title: 'Star Wars: Episode VII - The Force Awakens',
                Type: 'movie',
            },
            {
                ID: 'tt2527336',
                Poster:
                    'https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg',
                Title: 'Star Wars: Episode VIII - The Last Jedi',
                Type: 'movie',
            }
        )
        await expect(
            moviesHandler(undefined, undefined)
        ).resolves.toMatchSnapshot()
    })

    test('throws on error from getMovies', async () => {
        getMovies.mockRejectedValue(
            new HTTPError('error', StatusCodes.NOT_FOUND)
        )
        await expect(moviesHandler(undefined, undefined)).rejects.toThrow(
            HTTPError
        )
    })
})
