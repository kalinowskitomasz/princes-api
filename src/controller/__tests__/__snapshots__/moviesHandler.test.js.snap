// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moviesHandler returns valid response 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": Object {
      "data": Array [
        123,
        10,
        32,
        32,
        34,
        73,
        68,
        34,
        58,
        32,
        34,
        116,
        116,
        50,
        52,
        56,
        56,
        52,
        57,
        54,
        34,
        44,
        10,
        32,
        32,
        34,
        80,
        111,
        115,
        116,
        101,
        114,
        34,
        58,
        32,
        34,
        104,
        116,
        116,
        112,
        115,
        58,
        47,
        47,
        109,
        46,
        109,
        101,
        100,
        105,
        97,
        45,
        97,
        109,
        97,
        122,
        111,
        110,
        46,
        99,
        111,
        109,
        47,
        105,
        109,
        97,
        103,
        101,
        115,
        47,
        77,
        47,
        77,
        86,
        53,
        66,
        79,
        84,
        65,
        122,
        79,
        68,
        69,
        122,
        78,
        68,
        65,
        122,
        77,
        108,
        53,
        66,
        77,
        108,
        53,
        66,
        97,
        110,
        66,
        110,
        88,
        107,
        70,
        116,
        90,
        84,
        103,
        119,
        77,
        68,
        85,
        49,
        77,
        84,
        103,
        122,
        78,
        122,
        69,
        64,
        46,
        95,
        86,
        49,
        95,
        83,
        88,
        51,
        48,
        48,
        46,
        106,
        112,
        103,
        34,
        44,
        10,
        32,
        32,
        34,
        84,
        105,
        116,
        108,
        101,
        34,
        58,
        32,
        34,
        83,
        116,
        97,
        114,
        32,
        87,
        97,
        114,
        115,
        58,
        32,
        69,
        112,
        105,
        115,
        111,
        100,
        101,
        32,
        86,
        73,
        73,
        32,
        45,
        32,
        84,
        104,
        101,
        32,
        70,
        111,
        114,
        99,
        101,
        32,
        65,
        119,
        97,
        107,
        101,
        110,
        115,
        34,
        44,
        10,
        32,
        32,
        34,
        84,
        121,
        112,
        101,
        34,
        58,
        32,
        34,
        109,
        111,
        118,
        105,
        101,
        34,
        10,
        125,
      ],
      "type": "Buffer",
    },
    "disturbed": false,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "content-type": Array [
          "application/json;charset=UTF-8",
        ],
      },
    },
    "status": 200,
    "statusText": "OK",
    "url": undefined,
  },
}
`;
